Gun websocket

dealloc opened the pull request:
Changes out websocket_client for the :gun library
This allows custom reconnection logic and handling close frames with the error code when Discord closes the gateway.

dealloc committed:
rework Websocket module

Swap out :websocket_client for :gun
Refactor the Websocket module to act like a regular GenServer
Override cowlib dependency to avoid conflicts with fake_server
Updated calls to Websocket.send to match new signature

dealloc committed:
convert gateway worker

Refactor the Wumpex.Gateway.Worker module to use the new Websocket format
(WIP) refactor the shard module to use new Websocket format
Re-enable the Shard startup

dealloc committed:
Implement reconnection logic

Implement reconnection strategies defined by the implementing module, and if no implementation is present shut down the server

dealloc committed:
implement reconnect logic

Implemented logic to handle reconnects
Refactored on_connected since it now only gets called on initial connection

dealloc committed:
move Wumpex.Gateway.Worker to Wumpex.Gateway

Refactor logic of the worker into the Gateway module
Updated Shard to use the new Gateway instead of Worker

dealloc committed:
auto implement optional callbacks

Remove conditional calling of optional callbacks
Implement optional callbacks and make them overridable

dealloc committed:
Update tests

Update Websocket unit tests
Remove Gateway tests since they can't be unit tested

lambstream commented on lib/wumpex/base/websocket.ex:
```diff
@@ -1,202 +1,345 @@
-  You can pass whether to use binary or text mode using the `options`:
-      # Will use text mode
-      Websocket.send(socket, "Hello world", mode: :text)
+  This funcion takes the current state and returns the new state for the process.
```
```suggestion
  This function takes the current state and returns the new state for the process.
```

lambstream commented on lib/wumpex/gateway.ex:
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
+  end
+
+  @impl Websocket
+  def on_connected(options) do
```
```suggestion
  @doc """
  Builds the gateway state
  """
  def on_connected(options) do
```
maybe add some docs here

dealloc replied:
Documentation is implied by the `@impl Websocket` directive, which marks this an implementation of the websocket behaviour (which is documented).

lambstream commented on lib/wumpex/gateway.ex:
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
+  end
+
+  @impl Websocket
+  def on_disconnected(state) do
```
Add documentation please

dealloc replied:
Documentation is implied by the `@impl Websocket` directive, which marks this an implementation of the websocket behaviour (which is documented).

lambstream commented on lib/wumpex/gateway.ex:
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
+  end
+
+  @impl Websocket
+  def on_reconnected(state) do
```
Add documentation please 😄

dealloc replied:
Documentation is implied by the `@impl Websocket` directive, which marks this an implementation of the websocket behaviour (which is documented).

lambstream commented on lib/wumpex/gateway.ex:
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
+  end
+
+  @impl Websocket
+  def handle_frame({:binary, etf}, state) do
```
Add documentation please 😄

dealloc replied:
Documentation is implied by the `@impl Websocket` directive, which marks this an implementation of the websocket behaviour (which is documented).

lambstream commented on lib/wumpex/gateway.ex:
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
+    {:noreply, %{state | ack: false}}
+  end
+
+  # Handles HELLO
```
Maybe more info on what HELLO is or reference to discord docs?

lambstream commented on lib/wumpex/gateway.ex:
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
+    %{state | sequence: sequence}
+  end
+
+  def dispatch(%{op: 0, s: sequence, t: :GUILD_CREATE, d: event}, %{guild_sup: guild_sup} = state) do
```
Add documentation please 😄

lambstream commented on lib/wumpex/gateway.ex:
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
+    %{state | sequence: sequence}
+  end
+
+  def dispatch(%{op: 0, s: sequence, t: :GUILD_DELETE, d: %{id: guild_id}}, state) do
```
Add documentation please 😄

lambstream commented on lib/wumpex/gateway.ex:
```diff
@@ -13,10 +64,171 @@ defmodule Wumpex.Gateway do
+    %{state | sequence: sequence}
+  end
+
+  def dispatch(%{op: 0, s: sequence, t: event_name, d: %{guild_id: guild_id} = event}, state) do
```
Add documentation please 😄

lambstream commented on test/support/websocket_helpers.exs:
```diff
@@ -45,3 +45,18 @@ defmodule WebsocketHelpers do
+  @moduledoc false
+  use Wumpex.Base.Websocket
+
+  def on_connected(options) do
```
Add documentation please 😄

lambstream commented on test/support/websocket_helpers.exs:
```diff
@@ -45,3 +45,18 @@ defmodule WebsocketHelpers do
+    Keyword.fetch!(options, :worker)
+  end
+
+  def handle_frame(frame, state) do
```
Add documentation please 😄

dealloc committed:
update documentation && formatting

Updated formatting of code
Updated missing documentation in Websocket and Gateway
